### ISSUE TITLE
curl exception against azurite: "Connection was closed by the server while trying to read a response"

### ISSUE THREAD

**Issue by ahojnnes** [Azure/azure-sdk-for-cpp#4206]
When executing blob storage requests against azurite, our code occasionally hits the following exception that is bubbled up from the `curl.cpp` wrapper: "Connection was closed by the server while trying to read a response" - it seems that there could be three places, where this exception may be thrown. I have not managed to assemble a short code snippet to reproduce the issue, and we only observe this when running our emulated service against azurite.

What I see in practice when this issue happens:
1. Many micro-services are talking in parallel to azurite. 
2. One of the services fails to communicate and freezes for roughly 30 seconds in a call to an azure-storage-blobs method. It eventually throws the exception above.
3. I can consistently reproduce the issue on a call that looks as follows:
```
auto response = blobServiceClient->GetBlobContainerClient(container).GetBlobClient(name).Download();
response.Value.BodyStream->ReadToCount(
    reinterpret_cast<uint8_t*>(val.data()), val.size());
```
but it also happens for other calls.
4. The corresponding request never shows up in the azurite logs, which might indicate in a (temp) failure to talk to azurite, etc.
5. I have never seen the same issue when running the same code against real Azure blob storage.

Please let me know, if I can help with more details. Unfortunately, I cannot share the actual source code to replicate the issue.

**Comment by RickWinter** [comment]
Are you able to reproduce this problem in a debugger.  If so can you capture the stacktrace and share the Azure SDK portions of the stack.  It would greatly help us in diagnosing the issue.

**Comment by ahojnnes** [comment]
This is as much as I can give you:
```
* thread #46, name = 'azlocopt_servic', stop reason = breakpoint 1.1
  * frame #0: 0x00007ffff4bb0ce0 libc++abi.so.1`__cxa_throw
    frame #1: 0x00007ffff63618ad libazure-core.so`Azure::Core::Http::CurlSession::ReadStatusLineAndHeadersFromRawResponse(Azure::Core::Context const&, bool) + 1917
    frame #2: 0x00007ffff6360919 libazure-core.so`Azure::Core::Http::CurlSession::Perform(Azure::Core::Context const&) + 1241
    frame #3: 0x00007ffff635d7dd libazure-core.so`Azure::Core::Http::CurlTransport::Send(Azure::Core::Http::Request&, Azure::Core::Context const&) + 269
    frame #4: 0x00007ffff63813a3 libazure-core.so`Azure::Core::Http::Policies::_internal::TransportPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Http::Policies::NextHttpPolicy, Azure::Core::Context const&) const + 51
    frame #5: 0x00007ffff637e056 libazure-core.so`Azure::Core::Http::Policies::NextHttpPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Context const&) + 54
    frame #6: 0x00007ffff637cf1d libazure-core.so`Azure::Core::Http::Policies::_internal::LogPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Http::Policies::NextHttpPolicy, Azure::Core::Context const&) const + 349
    frame #7: 0x00007ffff637e056 libazure-core.so`Azure::Core::Http::Policies::NextHttpPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Context const&) + 54
    frame #8: 0x00007ffff637ed0a libazure-core.so`Azure::Core::Http::Policies::_internal::RequestActivityPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Http::Policies::NextHttpPolicy, Azure::Core::Context const&) const + 602
    frame #9: 0x00007ffff637e056 libazure-core.so`Azure::Core::Http::Policies::NextHttpPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Context const&) + 54
    frame #10: 0x00007ffff64278af libazure-storage-blobs.so`Azure::Storage::_internal::SharedKeyPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Http::Policies::NextHttpPolicy, Azure::Core::Context const&) const + 319
    frame #11: 0x00007ffff637e056 libazure-core.so`Azure::Core::Http::Policies::NextHttpPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Context const&) + 54
    frame #12: 0x00007ffff63bc3f0 libazure-storage-common.so`Azure::Storage::_internal::StoragePerRetryPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Http::Policies::NextHttpPolicy, Azure::Core::Context const&) const + 688
    frame #13: 0x00007ffff637e056 libazure-core.so`Azure::Core::Http::Policies::NextHttpPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Context const&) + 54
    frame #14: 0x00007ffff63bc8bb libazure-storage-common.so`Azure::Storage::_internal::StorageSwitchToSecondaryPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Http::Policies::NextHttpPolicy, Azure::Core::Context const&) const + 555
    frame #15: 0x00007ffff637e056 libazure-core.so`Azure::Core::Http::Policies::NextHttpPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Context const&) + 54
    frame #16: 0x00007ffff637fd82 libazure-core.so`Azure::Core::Http::Policies::_internal::RetryPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Http::Policies::NextHttpPolicy, Azure::Core::Context const&) const + 402
    frame #17: 0x00007ffff637e056 libazure-core.so`Azure::Core::Http::Policies::NextHttpPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Context const&) + 54
    frame #18: 0x00007ffff6381241 libazure-core.so`Azure::Core::Http::Policies::_internal::TelemetryPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Http::Policies::NextHttpPolicy, Azure::Core::Context const&) const + 113
    frame #19: 0x00007ffff637e056 libazure-core.so`Azure::Core::Http::Policies::NextHttpPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Context const&) + 54
    frame #20: 0x00007ffff6411268 libazure-storage-blobs.so`Azure::Core::Http::Policies::_internal::RequestIdPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Http::Policies::NextHttpPolicy, Azure::Core::Context const&) const + 152
    frame #21: 0x00007ffff637e056 libazure-core.so`Azure::Core::Http::Policies::NextHttpPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Context const&) + 54
    frame #22: 0x00007ffff6427b10 libazure-storage-blobs.so`Azure::Storage::_internal::StorageServiceVersionPolicy::Send(Azure::Core::Http::Request&, Azure::Core::Http::Policies::NextHttpPolicy, Azure::Core::Context const&) const + 144
    frame #23: 0x00007ffff64a2d8b libazure-storage-blobs.so`Azure::Storage::Blobs::_detail::BlobClient::Download(Azure::Core::Http::_internal::HttpPipeline&, Azure::Core::Url const&, Azure::Storage::Blobs::_detail::BlobClient::DownloadBlobOptions const&, Azure::Core::Context const&) + 2347
    frame #24: 0x00007ffff641884b libazure-storage-blobs.so`Azure::Storage::Blobs::BlobClient::Download(Azure::Storage::Blobs::DownloadBlobOptions const&, Azure::Core::Context const&) const + 1803
```

**Comment by ahojnnes** [comment]
(stacktrace for latest versions in vcpkg)

**Comment by ahojnnes** [comment]
Also got you the verbose logging output around the impacted area:
```
[2022-12-20T21:40:42.4036110Z] INFO  : HTTP Request : GET http://127.0.0.1:30300/devstoreaccount1/container/blob_name
authorization : REDACTED
user-agent : azsdk-cpp-storage-blobs/12.6.2 (Linux 5.15.79.1-microsoft-standard-WSL2 x86_64 #1 SMP Wed Nov 23 01:01:46 UTC 2022)
x-ms-client-request-id : fe1cae15-abb9-498b-72f1-e9fa0e9430d6
x-ms-date : REDACTED
x-ms-version : REDACTED
[2022-12-20T21:40:42.4037370Z] DEBUG : [CURL Transport Adapter]: Creating a new session.
[2022-12-20T21:40:42.4038610Z] DEBUG : [CURL Transport Adapter]: Re-using connection from the pool.
[2022-12-20T21:40:42.4039460Z] DEBUG : [CURL Transport Adapter]: No Host in request headers. Adding it
[2022-12-20T21:40:42.4040260Z] DEBUG : [CURL Transport Adapter]: No content-length in headers. Adding it
[2022-12-20T21:40:42.4041460Z] DEBUG : [CURL Transport Adapter]: Send request without payload
[2022-12-20T21:40:42.4045050Z] DEBUG : [CURL Transport Adapter]: Parse server response
[2022-12-20T21:40:42.4050340Z] WARN  : HTTP Transport error: Connection was closed by the server while trying to read a response
[2022-12-20T21:40:42.4050680Z] INFO  : HTTP Retry attempt #1 will be made in 4900ms.
[2022-12-20T21:40:47.3055940Z] INFO  : HTTP Request : GET http://127.0.0.1:30300/devstoreaccount1/container/blob_name
authorization : REDACTED
user-agent : azsdk-cpp-storage-blobs/12.6.2 (Linux 5.15.79.1-microsoft-standard-WSL2 x86_64 #1 SMP Wed Nov 23 01:01:46 UTC 2022)
x-ms-client-request-id : fe1cae15-abb9-498b-72f1-e9fa0e9430d6
x-ms-date : REDACTED
x-ms-version : REDACTED
[2022-12-20T21:40:47.3057630Z] DEBUG : [CURL Transport Adapter]: Creating a new session.
[2022-12-20T21:40:47.3059160Z] DEBUG : [CURL Transport Adapter]: Re-using connection from the pool.
[2022-12-20T21:40:47.3060690Z] DEBUG : [CURL Transport Adapter]: No Host in request headers. Adding it
[2022-12-20T21:40:47.3062130Z] DEBUG : [CURL Transport Adapter]: No content-length in headers. Adding it
[2022-12-20T21:40:47.3063660Z] DEBUG : [CURL Transport Adapter]: Send request without payload
[2022-12-20T21:40:47.3066540Z] DEBUG : [CURL Transport Adapter]: Parse server response
[2022-12-20T21:40:47.3073130Z] WARN  : HTTP Transport error: Connection was closed by the server while trying to read a response
[2022-12-20T21:40:47.3073640Z] INFO  : HTTP Retry attempt #2 will be made in 6836ms.
[2022-12-20T21:40:54.1439740Z] INFO  : HTTP Request : GET http://127.0.0.1:30300/devstoreaccount1/container/blob_name
authorization : REDACTED
user-agent : azsdk-cpp-storage-blobs/12.6.2 (Linux 5.15.79.1-microsoft-standard-WSL2 x86_64 #1 SMP Wed Nov 23 01:01:46 UTC 2022)
x-ms-client-request-id : fe1cae15-abb9-498b-72f1-e9fa0e9430d6
x-ms-date : REDACTED
x-ms-version : REDACTED
[2022-12-20T21:40:54.1440620Z] DEBUG : [CURL Transport Adapter]: Creating a new session.
[2022-12-20T21:40:54.1442090Z] DEBUG : [CURL Transport Adapter]: Re-using connection from the pool.
[2022-12-20T21:40:54.1443550Z] DEBUG : [CURL Transport Adapter]: No Host in request headers. Adding it
[2022-12-20T21:40:54.1445370Z] DEBUG : [CURL Transport Adapter]: No content-length in headers. Adding it
[2022-12-20T21:40:54.1446440Z] DEBUG : [CURL Transport Adapter]: Send request without payload
[2022-12-20T21:40:54.1448880Z] DEBUG : [CURL Transport Adapter]: Parse server response
[2022-12-20T21:40:54.1454490Z] WARN  : HTTP Transport error: Connection was closed by the server while trying to read a response
[2022-12-20T21:40:54.1455280Z] INFO  : HTTP Retry attempt #3 will be made in 20088ms.
[2022-12-20T21:41:14.2338810Z] INFO  : HTTP Request : GET http://127.0.0.1:30300/devstoreaccount1/container/blob_name
authorization : REDACTED
user-agent : azsdk-cpp-storage-blobs/12.6.2 (Linux 5.15.79.1-microsoft-standard-WSL2 x86_64 #1 SMP Wed Nov 23 01:01:46 UTC 2022)
x-ms-client-request-id : fe1cae15-abb9-498b-72f1-e9fa0e9430d6
x-ms-date : REDACTED
x-ms-version : REDACTED
[2022-12-20T21:41:14.2340270Z] DEBUG : [CURL Transport Adapter]: Creating a new session.
[2022-12-20T21:41:14.2342310Z] DEBUG : [CURL Transport Adapter]: Re-using connection from the pool.
[2022-12-20T21:41:14.2343830Z] DEBUG : [CURL Transport Adapter]: No Host in request headers. Adding it
[2022-12-20T21:41:14.2344890Z] DEBUG : [CURL Transport Adapter]: No content-length in headers. Adding it
[2022-12-20T21:41:14.2346920Z] DEBUG : [CURL Transport Adapter]: Send request without payload
[2022-12-20T21:41:14.2350800Z] DEBUG : [CURL Transport Adapter]: Parse server response
[2022-12-20T21:41:14.2356400Z] WARN  : HTTP Transport error: Connection was closed by the server while trying to read a response
```

**Comment by Jinming-Hu** [comment]
Azurite is designed to be an emulator of Azure Storage, but there are many, many subtle nuances of behavior from the real Azure storage service.
Because libcurl transport layer is **not** a general-purpose HTTP client, it cannot handle all of the specifics of HTTP protocol correctly (we've met such issues a few times). 
If this issue can only be reproduced with Azurite, it's likely to be caused by one of these nuances. (Both Azurite and Azure Storage return valid HTTP response, but libcurl transport layer can only handle the one from Azure storage.)
Being compatible with Azurite is not our goal. We can still look into this issue, but it's anyway not our priority.

**Comment by ahojnnes** [comment]
Thanks for your response and the explanation. It is unfortunate that this is not a goal, as Azurite does provide a convenient way to enable automated testing in a fully emulated environment (e.g. in CI) without the need to connect to a real blob storage account. However, if this is not a stated goal of this project and not tested as part of your CI setup, then this seems like the best path to hit the same kind of issues for us in the future.

What is your suggested setup to enable automated testing? Should one connect to real Azure storage or are there alternatives?

**Comment by Jinming-Hu** [comment]
@ahojnnes I'll discuss with team to see if we can prioritize this item.

**Comment by ahojnnes** [comment]
Please see: https://github.com/Azure/azure-sdk-for-cpp/pull/4208

**Comment by Jinming-Hu** [comment]
Hi @ahojnnes , what is the root cause of this problem? Is it because Azurite closes all the connections without letting client know? Then client tries to reuse the connections in pool, which are all closed by Azurite. Each attempt ends up with an exception and when the number of attempts exceed `MaxRetries`, the request fails. Is that the reason?

**Comment by ahojnnes** [comment]
> Is it because Azurite closes all the connections without letting client know? Then client tries to reuse the connections in pool, which are all closed by Azurite. Each attempt ends up with an exception and when the number of attempts exceed MaxRetries, the request fails. Is that the reason?

Yes, that's my understanding of what is happening. In any case, it will be good to be robust on the SDK side to this scenario. The server not properly closing the connection can happen for any number of reasons and will lead to the same issues.

**Comment by Jinming-Hu** [comment]
@ahojnnes with your fix in #4208, the request will eventually succeed, how about the 30-second delay? Is it able to get rid of the delay as well?

**Comment by antkmsft** [comment]
Fixed in #4208.
